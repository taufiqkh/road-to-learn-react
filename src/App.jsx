import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';

const REPO_LIST = [
  {
    title: 'React',
    url: 'https://facebook.github.io/react/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
  },
  {
    title: 'Redux',
    url: 'https://github.com/reactjs/redux',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
  },
];

function isSearched(query) {
  return function isQuerySearchedOn(item) {
    return !query || item.title.toLowerCase().indexOf(query.toLowerCase()) !== -1;
  };
}

const Search = ({ value, onChange, children }) =>
  (
    <form>
      {children} <input type="text" value={value} onChange={onChange} />
    </form>
  );

const Table = ({ list, pattern }) =>
  (
    <div>
      { list.filter(isSearched(pattern)).map(item =>
        <div key={item.objectID}>
          <span><a href={item.url}>{item.title}</a></span>
          <span>{item.author}</span>
          <span>{item.numComments}</span>
          <span>{item.points}</span>
        </div>,
      )}
    </div>
  );

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      list: REPO_LIST,
      query: '',
    };

    this.onSearchChange = this.onSearchChange.bind(this);
  }

  onSearchChange(event) {
    this.setState({ query: event.target.value });
  }

  render() {
    const { query, list } = this.state;

    return (
      <div className="App">
        <Search value={query} onChange={this.onSearchChange}>
          Search
        </Search>
        <Table list={list} pattern={query} />
        <div className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
        </div>
        <p className="App-intro">
          To get started, edit <code>src/App.js</code> and save to reload.
        </p>
      </div>
    );
  }
}

export default App;
