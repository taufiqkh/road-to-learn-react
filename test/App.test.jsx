import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import App from '../src/App';
import { Search, Button, Table, Loading } from '../src/App';

describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
  });

  test('snapshots', () => {
    const component = renderer.create(
      <App />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Search', () => {
  it('renders', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Search>Search</Search>, div);
  });

  test('snapshots', () => {
    const component = renderer.create(
      <Search>Search</Search>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Button', () => {
  it('renders', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Button>Give me more</Button>, div);
  });

  test('snapshot', () => {
    const component = renderer.create(
      <Button>Give me more</Button>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Table', () => {
  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
      { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' },
    ],
  };

  it('renders', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
  });

  it('shows two items in lists', () => {
    const element = shallow(
      <Table {...props} />
    );
    expect(element.find('.table-row').length).toBe(2);
  });

  test('snapshot', () => {
    const component = renderer.create(
      <Table {...props} />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Loading', () => {
  it('renders', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Loading />, div);
  });

  test('snapshot', () => {
    const component = renderer.create(
      <Loading />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })
});