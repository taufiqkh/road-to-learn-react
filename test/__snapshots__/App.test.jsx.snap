exports[`App snapshots 1`] = `
<div
  className="page">
  <div
    className="interactions">
    <form
      onSubmit={[Function]}>
      <input
        onChange={[Function]}
        type="text"
        value="redux" />
      <button
        type="submit">
        Search
      </button>
    </form>
  </div>
  <div
    className="table">
    <div
      className="table-header">
      <span
        style={
          Object {
            "width": "40%",
          }
        }>
        <button
          className="button-inline"
          onClick={[Function]}
          type="button">
          Title
        </button>
      </span>
      <span
        style={
          Object {
            "width": "30%",
          }
        }>
        <button
          className="button-inline"
          onClick={[Function]}
          type="button">
          Author
        </button>
      </span>
      <span
        style={
          Object {
            "width": "15%",
          }
        }>
        <button
          className="button-inline"
          onClick={[Function]}
          type="button">
          Comments
        </button>
      </span>
      <span
        style={
          Object {
            "width": "15%",
          }
        }>
        <button
          className="button-inline"
          onClick={[Function]}
          type="button">
          Points
        </button>
      </span>
    </div>
  </div>
  <div
    className="interactions">
    <div>
      Loading...
    </div>
  </div>
</div>
`;

exports[`Button snapshot 1`] = `
<button
  onClick={undefined}
  type="button">
  Give me more
</button>
`;

exports[`ButtonWithLoading snapshot isLoading 1`] = `
<div>
  Loading...
</div>
`;

exports[`ButtonWithLoading snapshot notLoading 1`] = `
<button
  onClick={undefined}
  type="button" />
`;

exports[`Loading snapshot 1`] = `
<div>
  Loading...
</div>
`;

exports[`Search snapshots 1`] = `
<form
  onSubmit={undefined}>
  <input
    onChange={undefined}
    type="text"
    value={undefined} />
  <button
    type="submit">
    Search
  </button>
</form>
`;

exports[`Table snapshot 1`] = `
<div
  className="table">
  <div
    className="table-header">
    <span
      style={
        Object {
          "width": "40%",
        }
      }>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button">
        Title
      </button>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button">
        Author
      </button>
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button">
        Comments
      </button>
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button">
        Points
      </button>
    </span>
  </div>
  <div
    className="table-row">
    <span
      style={
        Object {
          "width": "40%",
        }
      }>
      <a
        href={undefined}>
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }>
      1
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }>
      1
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }>
      2
    </span>
  </div>
  <div
    className="table-row">
    <span
      style={
        Object {
          "width": "40%",
        }
      }>
      <a
        href={undefined}>
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }>
      2
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }>
      1
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }>
      2
    </span>
  </div>
</div>
`;
